### PR TITLE
Timx 287 ead field method refactor 1

### PR DESCRIPTION
**Pull request by jonavellecuerdo**
### Purpose and background context

Field method refactor for transform class Ead (Part 1).

* Added private"methods for key elements in EAD-formatted XML records (`<archdesc>`, `<archdesc><did>`, `<archdesc><controlaccess>`
* Added field methods and corresponding unit tests for `alternate_titles`, `citation`, `content_type`

**Note:** For the time-being, I've added calls to the private methods at the beginning of `get_optional_fields()`, which is required by field derivations that **will eventually be moved into their own field methods**.

### How can a reviewer manually see the effects of these changes?

1. Run `make test` and verify all unit tests are passing.
2. Run CLI command
   ```
   pipenv run transform -i tests/fixtures/ead/ead_record_all_fields.xml -o output/ead-transformed-records.json -s aspace
   ```
   Output: 
   ```
   2024-05-30 12:33:59,914 INFO transmogrifier.cli.main(): Logger 'root' configured with level=INFO
   2024-05-30 12:33:59,914 INFO transmogrifier.cli.main(): No Sentry DSN found, exceptions will not be sent to Sentry
   2024-05-30 12:33:59,914 INFO transmogrifier.cli.main(): Running transform for source aspace
   2024-05-30 12:33:59,930 WARNING transmogrifier.sources.transformer.get_valid_title(): Record repositories/2/resources/1 has multiple titles. Using the first title from the following titles found: ['Charles J. Connick Stained Glass Foundation Collection', 'Title 2', 'Title 3']
   2024-05-30 12:33:59,930 INFO transmogrifier.cli.main(): Completed transform, total records processed: 1, transformed records: 1, skipped records: 0, deleted records: 0
   2024-05-30 12:33:59,930 INFO transmogrifier.cli.main(): Total time to complete transform: 0:00:00.016147

   ```
### Includes new or updated dependencies?
NO

### Changes expectations for external applications?
NO

### What are the relevant tickets?
- https://mitlibraries.atlassian.net/browse/TIMX-287

### Developer
- [ ] All new ENV is documented in README
- [ ] All new ENV has been added to staging and production environments
- [ ] All related Jira tickets are linked in commit message(s)
- [ ] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed and verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: tests/sources/xml/test_ead.py [hunk 1]
```diff
@@ -1,10 +1,61 @@
+# ruff: noqa: E501
```
Comment (ehanson8) [inline tests/sources/xml/test_ead.py:1]: Perhaps just a personal opinion but I think it's preferable to not ignore E501 for a module (or even a line really). For the `xml_string` var in `create_ead_source_record_stub`, you can use spaces instead of tabs and you can break up the long citation strings like this:

```
assert Ead.get_citation(source_record) == (
        "Charles J. Connick Stained Glass Foundation Collection, "
        "VC-0002, box X. Massachusetts Institute of Technology, "
        "Department of Distinctive Collections, Cambridge, Massachusetts."
    )
```

File: tests/sources/xml/test_ead.py [hunk 2]
```diff
@@ -501,12 +563,11 @@ def test_ead_record_invalid_date_and_date_range_are_omitted(caplog):
             assert date.range.lte != "1989"
             assert date.range.gte != "2001"
             assert date.range.lte != "1999"
-
     assert ("has a date that couldn't be parsed: 'undated'") in caplog.text
     assert ("has a later start date than end date: '2001', '1999'") in caplog.text
 
 
-def test_ead_record_correct_identifiers_from_multiple_unitid(caplog):
+def test_ead_transform_with_multiple_unitid_gets_valid_ids():
```
Comment (ehanson8) [inline tests/sources/xml/test_ead.py:570]: `Ead` has more of these edge case tests than the other sources we've done so far, we might consider redoing these with stub records but not sure if that's worth the effort

File: tests/sources/xml/test_ead.py [hunk 3]
```diff
@@ -515,19 +576,223 @@ def test_ead_record_correct_identifiers_from_multiple_unitid(caplog):
...
+                        <emph>
+                            Foundation
+                            <emph>Collection</emph>
+                        </emph>
+                        <num>VC.0002</num>
+                    </unittitle>
+                    <unittitle></unittitle>
+                    <unittitle></unittitle>
+                </did>
+            </archdesc>
```
Comment (ehanson8) [inline tests/sources/xml/test_ead.py:626]: This is duplicated a lot,  you might refactor so top level elements like `<archdesc>` aren't repeated in every test

**Review by ghukill — CHANGES_REQUESTED** [review]
Overall, I think it looks great!   Left a request for docstrings in the helper methods, and a question, but overall looking good to me.

File: tests/sources/xml/test_ead.py [hunk 4]
```diff
@@ -515,19 +576,223 @@ def test_ead_record_correct_identifiers_from_multiple_unitid(caplog):
...
+        metadata_insert=(
+            """
+            <archdesc level="collection"></archdesc>
+            """
+        )
+    )
+    assert Ead.get_citation(source_record) is None
+
+
+def test_get_main_titles_success():
```
Comment (ehanson8) [inline tests/sources/xml/test_ead.py:697]: This would be easier to read if the order matched the method order in `ead.py` so I would move the `get_main_titles` tests further down

File: transmogrifier/sources/xml/ead.py [hunk 5]
```diff
@@ -362,19 +327,86 @@ def generate_name_identifier_url(cls, name_element: Tag) -> list | None:
...
+    def _get_collection_description(cls, source_record: Tag) -> Tag:
+        if collection_description := source_record.metadata.find(
+            "archdesc", level="collection"
+        ):
+            return collection_description
+        message = (
+            "Record skipped because key information is missing: "
+            '<archdesc level="collection">.'
+        )
+        raise SkippedRecordEvent(message)
```
Comment (ehanson8) [inline transmogrifier/sources/xml/ead.py:339]: Perfect use of `SkippedRecordEvent`!

File: transmogrifier/sources/xml/ead.py [hunk 6]
```diff
@@ -17,72 +18,36 @@
...
-            pass
-        else:
-            message = (
-                f"Record ID {self.get_source_record_id(xml)} is missing archdesc element"
-            )
-            logger.error(message)
-            return None
+        # <archdesc> and <did> elements are required when deriving optional fields
+        collection_description = self._get_collection_description(source_record)
+        collection_description_did = self._get_collection_description_did(source_record)
```
Comment (ehanson8) [inline transmogrifier/sources/xml/ead.py:36]: Wondering if these calls are still necessary up top, it seems they would raise the `SkippedRecordEvent` whenever they are called by the field methods

File: transmogrifier/sources/xml/ead.py [hunk 7]
```diff
@@ -362,19 +327,86 @@ def generate_name_identifier_url(cls, name_element: Tag) -> list | None:
...
+        ] or None
+
+    @classmethod
+    def get_citation(cls, source_record: Tag) -> str | None:
+        collection_description = cls._get_collection_description(source_record)
+        if (
+            citation_element := collection_description.find("prefercite", recursive=False)
+        ) and (
+            citation := cls.create_string_from_mixed_value(
+                citation_element, " ", ["head"]
```
Comment (ehanson8) [inline transmogrifier/sources/xml/ead.py:372]: We may want to start using named args for this

**Review by ehanson8 — COMMENTED** [review]
Very good, a few suggestions!

File: transmogrifier/sources/xml/ead.py (hunk 6 above)
Comment (ghukill) [reply transmogrifier/sources/xml/ead.py:36]: I left a comment that might be related: my assumption is that these will be removed once _fully_ refactored to field methods?

Comment (ehanson8) [reply transmogrifier/sources/xml/ead.py:36]: Ah, didn't see your comments before my review but that makes sense as a plan!

[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply transmogrifier/sources/xml/ead.py:36]: Ah, yes! I meant to write a note after marking the PR ready for review but forgot. 😅 Thank you for remembering! I added a note in the PR description!

Comment (jonavellecuerdo) [reply transmogrifier/sources/xml/ead.py:36]: See [response](https://github.com/MITLibraries/transmogrifier/pull/185#discussion_r1621324261); PR description updated with note. Sorry for the confusion!

File: transmogrifier/sources/xml/ead.py (hunk 7 above)
Comment (jonavellecuerdo) [reply transmogrifier/sources/xml/ead.py:372]: I was considering this but felt conflicted [with @ghukill 's comment on a previous PR](https://github.com/MITLibraries/transmogrifier/pull/177#discussion_r1613894603), but perhaps the guideline @ghukill is only applicable for "methods with single positional arguments" or "when testing"?

In any case, I will update to use named args! I think `create_string_from_mixed_value` is nifty but is difficult to understand just by the name alone.

Comment (ghukill) [reply transmogrifier/sources/xml/ead.py:372]: To clarify my comment a bit...

If I'm reading correctly, this is the signature:
```python
def create_string_from_mixed_value(
        cls,
        xml_element: Tag,
        separator: str = "",
        skipped_elements: list[str] | None = None,
    )...
```

So anytime called, I'd expect to see:
  - required, positional argument `xml_element` is not named, but just passed positionally
  - the rest of them, passed as name arguments, because named in the signature
    - _unless_ the defaults are okay, and then okay to omit

Does that mesh with the discussion here?

Comment (ghukill) [reply transmogrifier/sources/xml/ead.py:372]: So in this case, I'd expect:

```python
cls.create_string_from_mixed_value(
    citation_element,
    separator = " ",
    skipped_elements = ["head"]
)
```

Open to pushback, but this convention feels like a pretty widely followed convention?

Comment (jonavellecuerdo) [reply transmogrifier/sources/xml/ead.py:372]: Hmm, I think my confusion was I took "named arguments" (or keyword arguments) literally. 🤔 Technically, all the arguments are positional since the signature isn't as follows:

```python
def create_string_from_mixed_value(
        cls,
        xml_element: Tag,
        *,
        separator: str = "",
        skipped_elements: list[str] | None = None,
)
```

I'm definitely on board with the proposed convention; just curious if it's more accurate to say: 

> if overriding default positional arguments, use keyword arguments in the method call
?

File: tests/sources/xml/test_ead.py (hunk 1 above)
Comment (jonavellecuerdo) [reply tests/sources/xml/test_ead.py:1]: Good idea! On a somewhat related note (and is also why I hesitated to reformat the strings):

When I save on an open XML file, I have a tool in my IDE that reformats the XML. It results in newlines being added throughout the XML file, which then causes the tests to fail. 

@ehanson8 @ghukill Do you think it's worth creating an issue to whitespace/newline-proof the XML transforms? 🤔

File: transmogrifier/sources/xml/ead.py (hunk 7 above)
Comment (ghukill) [reply transmogrifier/sources/xml/ead.py]: If I'm reading it right, this is the actual function definition:

```python
@classmethod
def create_string_from_mixed_value(
    cls,
    xml_element: Tag,
    separator: str = "",
    skipped_elements: list[str] | None = None,
) -> str:
   ...
```

Apologies if I'm being imprecise with my words here or earlier, but I would characterize the arguments like:
  * `cls`: required, placeholder, can kind of ignore
  * `xml_element`: required, positional
  * `separator`: because it has equals sign `=`, it's named, optional, with default
  * `skipped_elements`: same as separator

I take your point that _technically_ all the arguments are positional, good point.  But I suppose my position is that _conventionally_ -- and this is open to debate for sure -- if an argument has an equals sign `=` than it's treated as a named argument, and therefore optional with a default value.

This [SO post](https://stackoverflow.com/a/57819001) breaks it down nicely, and exposes that I _have_ been lazy with my terminology.

If we ever do deal with functions that use `*args` or `**kwargs`, then I think it gets more subtle (to your example above), but for most signatures, I feel like it's often positional (no defaults) and then named/keyword (with defaults).  My vote is to
   * a) _not_ use the arg name for positionals
   * b) use the arg name for named/keyword

Thoughts?  or am I just making a mountain out of a mole hill 😅?

Comment (jonavellecuerdo) [reply transmogrifier/sources/xml/ead.py]: Thank you for clarifying! I accept / understand your definitions and we are on the same page! I do wonder whether we should start using `*` or `/` in our function signatures to more explicitly distinguish positional vs named/keyword arguments but beyond the scope of this PR. 🤓 If I remember correctly, [Ruff throws errors for boolean positional arguments (FBT001)](https://docs.astral.sh/ruff/rules/boolean-type-hint-positional-argument/).

Circling back to original comment: [changes have been made](https://github.com/MITLibraries/transmogrifier/pull/185/commits/de46532bbe2390a48aed2927793e60a130ba0a50).

Comment (ehanson8) [reply transmogrifier/sources/xml/ead.py]: That feels right to me but I honestly am learning some of these conventions in these PR conversations 🙃

File: tests/sources/xml/test_ead.py (hunk 1 above)
Comment (ghukill) [reply tests/sources/xml/test_ead.py:1]: We'd probably want to preserve whitespace inside an element (its text), as this is often intentional by the record creator.  I think it would fundamentally change the nature of the content if we introduce newline characters inside a long string of text for an element.

But reformatting the XML in a way that just moves elements around, but not element text, that seems okay / good.

Comment (jonavellecuerdo) [reply tests/sources/xml/test_ead.py:1]: I think I'll forego creating an issue for now as the newlines are resulting from me not turning off the XML formatter in my IDE 😅 . 

Circling back to original comment, I removed the syntax to ignore `E501` errors and made updates accordingly!

File: transmogrifier/sources/xml/ead.py (hunk 7 above)
Comment (ghukill) [reply transmogrifier/sources/xml/ead.py]: This conversation is the first time I've seen the `/` operator for function/method signatures, and helped clarify for me what the `*` operator does _without_ an immediate following variable name like `*args`.  

This helped me understand it more tacitly:

```python
In [1]: def foo(a, b, *, c="horse"):
   ...:     print(a,b,c)
   ...: 

In [2]: foo(42, 101, 999, 888, c='zebra')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 foo(42, 101, 999, 888, c='zebra')

TypeError: foo() takes 2 positional arguments but 4 positional arguments (and 1 keyword-only argument) were given

In [3]: def bar(a, b, *args, c="horse"):
   ...:     print(a,b,args,c)
   ...: 

In [4]: bar(42, 101, 999, 888, c='zebra')
42 101 (999, 888) zebra
```

I'm definitely not opposed to using `*` and `/` in signatures if we find them useful.  Here is an edge case where I could see it being useful, when you want to force keyword argument or no positional:

```python
def baz(*, a=1):
    pass

In [6]: baz()
1

In [7]: baz(42)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[7], line 1
----> 1 baz(42)

TypeError: baz() takes 0 positional arguments but 1 was given

In [8]: baz(a=101)
101    
```

Ineresting stuff!  In cases without `*` or `/`, glad we're honing in on some conventions, and curious if we ever run into situations where it would be nice to use them.  I read an article this morning that I found pretty convincing that the main appeal is often APIs, where you want to finely control how functions are called by users potentially not familiar with the library.

File: tests/sources/xml/test_ead.py (hunk 2 above)
Comment (jonavellecuerdo) [reply tests/sources/xml/test_ead.py:572]: @ehanson8 For this PR, I'll stick to just renaming the test and will revisit when I work on the `get_identifiers` field method.

File: tests/sources/xml/test_ead.py (hunk 4 above)
Comment (jonavellecuerdo) [reply tests/sources/xml/test_ead.py]: Updated!

File: tests/sources/xml/test_ead.py (hunk 3 above)
Comment (jonavellecuerdo) [reply tests/sources/xml/test_ead.py]: @ehanson8 @ghukill Let me know what you think of the updates I made to the [`create_ead_source_record_stub` method](https://github.com/MITLibraries/transmogrifier/pull/185/commits/b8adf19271c53d9f2b0f49d50d8b71ed0c68bd20#diff-4e446208d5381af3faa605c03e15486e64e9af30c4d54e8f40bec821cb87c851R11-R74). It was the only way I could think of simplifying some of the `metadata_insert` arguments throughout the testing module (removes the `<archdesc>` and `<archdesc><did>` elements from the string.

In general, we want to avoid complex functions in the testing module, but maybe this is a simple enough method that it isn't a problem? 🤔

File: transmogrifier/sources/xml/ead.py [hunk 8]
```diff
@@ -362,19 +327,86 @@ def generate_name_identifier_url(cls, name_element: Tag) -> list | None:
         return None
 
     @classmethod
-    def get_main_titles(cls, xml: Tag) -> list[str]:
+    def _get_collection_description(cls, source_record: Tag) -> Tag:
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply transmogrifier/sources/xml/ead.py:332]: Docstrings have been added!

File: transmogrifier/sources/xml/ead.py [hunk 9]
```diff
@@ -362,19 +327,86 @@ def generate_name_identifier_url(cls, name_element: Tag) -> list | None:
...
+        ):
+            return collection_description
+        message = (
+            "Record skipped because key information is missing: "
+            '<archdesc level="collection">.'
+        )
+        raise SkippedRecordEvent(message)
+
+    @classmethod
+    def _get_collection_description_did(cls, source_record: Tag) -> Tag:
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply transmogrifier/sources/xml/ead.py:354]: Docstrings have been added!

File: tests/sources/xml/test_ead.py (hunk 2 above)
Comment (ehanson8) [reply tests/sources/xml/test_ead.py:589]: That's totally fine, just food for thought!

File: tests/sources/xml/test_ead.py [hunk 10]
```diff
@@ -9,29 +8,27 @@
 
 
 def create_ead_source_record_stub(
-    header_insert: str = "", metadata_insert: str = ""
+    header_insert: str = "",
+    metadata_insert: str = "",
+    parent_element: Literal["archdesc", "did"] = None,  # noqa: RUF013
```
Comment (ehanson8) [inline tests/sources/xml/test_ead.py:13]: This isn't too complex in my mind and saves space in the unit tests but I defer to @ghukill for final approval of this

**Review by ehanson8 — APPROVED** [review]
Looks good to me and I'm glad this PR led to some learning for all of us! 🙂

File: transmogrifier/sources/xml/ead.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ghukill) [reply transmogrifier/sources/xml/ead.py:332]: Thanks!

File: transmogrifier/sources/xml/ead.py [hunk 11]
```diff
@@ -362,38 +329,139 @@ def generate_name_identifier_url(cls, name_element: Tag) -> list | None:
...
     @classmethod
-    def get_main_titles(cls, xml: Tag) -> list[str]:
+    def _get_collection_description(cls, source_record: Tag) -> Tag:
+        """Get element with archival description for a collection.
+
+        For EAD-formatted XML, all of the information about a collection
+        that is relevant to the TIMDEX data model is
+        encapsulated within the <archdesc level="collection"> element.
+        If this element is missing, it suggests that there is a structural
+        error in the record.
```
Comment (ghukill) [inline transmogrifier/sources/xml/ead.py:339]: Perfect, thanks.

**Review by ghukill — APPROVED** [review]
Looks good to me!

Thanks again for good discussions.  I'm glad in these early updates, we're hashing out some of these conventions and details.
